### ISSUE TITLE
cuda out of memory

### ISSUE THREAD

**Issue by YJonmo** [facebookresearch/silk#6]
Thanks for this great work.

When I set the 'silk-inference.py' to use dense descriptors, it goes out of memory:
model = get_model(default_outputs=("dense_positions", "dense_descriptors"))

The images are of size 256^2
With NVIDIA GeForce RTX 3090: 
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  Off  | 00000000:00:1B.0 Off |                  N/A |
| 30%   32C    P8    22W / 350W |      8MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

The reason that I am trying to get the dense descriptor running is to get more key points from areas with low textures. I made a simple comparison between LoFTR and SiLK (sparse) and as you can see the LoFTR provides more key points for low-texture areas:

SiLK (sparse):
![image](https://user-images.githubusercontent.com/12040950/233507454-8d5efd7c-beb8-4c1d-939a-c8f620c5bc4e.png)

and LoFTR for the same image:
![image](https://user-images.githubusercontent.com/12040950/233507551-04218a34-ce08-4e77-acf5-d7369d16b5ad.png)

In other words, how can I get a homogenous pattern of key points from SiLK similar to LoFTR.

**Comment by mertkaraoglu** [comment]
Hi YJonmo,

By nature, SILK is a keypoint (extraction and) matching method, unlike LoFTR which is a (semi-) dense matching method. LoFTR computes the dense features at the 1/8 resolution of the input image. which gives a much smaller memory foot-print in comparison to SILK (which's features space is at 1/4th of the input resolution).

One straight-forward idea is, you select every second (or fourth) pixel in the feature space of the source image and therefore end up with a similar grid-like sampling. However, to keep the accuracy high, ensure your search space on the target image is still every pixel of the feature space.

**Comment by gleize** [comment]
Hi @YJonmo,

Thanks @mertkaraoglu for the reply. I just want to correct one thing.

> [...] SILK (which's features space is at 1/4th of the input resolution).

This is not true. SiLK's feature space is at the original input resolution. This is caused by the removal of the `max-pooling` layer from the original VGG backbone.

This means that when @YJonmo runs the matching algorithm on dense keypoints, the similarity matrix has a size of $(238^2)\times(238^2)$ (238 being 256 minus the unpadded convolutions "eating" the border). This is quite large and I suspect this is where the memory issue comes from.

@YJonmo here are a few options :
* Feature subsampling as @mertkaraoglu kindly suggested.
* Downsizing image before running SiLK.
* Re-write the similarity matrix computation to iterate in smaller "blocks" (I might add that one if there is popular demand for it).

There are pro and cons to each option, but essentially it will depend on how accurate / fast you need it to be.

**Comment by mertkaraoglu** [comment]
> 

Thanks for the correction, that part is also clearer for me now.

**Comment by YJonmo** [comment]
Thanks a lot for the feedbacks. I will try your suggestions.

**Comment by YJonmo** [comment]
I have a question that might be irrelevant to this thread:
My camera is a lens camera and the zoom change will result in intrinsic change. Can I still use this equation to estimate the homogrpahy between two frames:

`estimated_homography, _ = cv2.findHomography(
        matched_points_0,
        matched_points_1,
        cv2.RANSAC    )`

**Comment by gleize** [comment]
@YJonmo 

The zoom should only change the scale, which can be handled by the homography. So I don't see any problem with it, as long as the homography is the right transformation for what you're trying to do (i.e. handling images where things are far away, or living on a planar surface).

**Comment by YJonmo** [comment]
thank you very much

**Comment by gleize** [comment]
Answers moved to FAQ. Closing now.
